components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt

```diff
@@ -24,15 +26,22 @@ typealias OnNeedToRequestPermissions = (session: Session, download: Download) ->
  * @property onNeedToRequestPermissions it will be call when you need to request permission,
  * before a download can be performed.
  * @property onDownloadCompleted a callback to be notified when a download is completed.
- * @property downloadManager a reference of the [DownloadManager] who is in charge of performing the downloads.
- * @property sessionManager a reference of the [SessionManager].
+ * @property downloadManager a reference of [DownloadManager] who is in charge of performing the downloads.
+ * @property sessionManager a reference of [SessionManager].
+ * @property fragmentManager a reference of [FragmentManager]. If you provide a [fragmentManager]
+ * a dialog will show before every download.
+ * @property fragmentManager a reference of [FragmentManager].
+ * @property dialog a reference of [DownloadDialogFragment]. If it is not provided, it will be
+ * initialized with an instance of [SimpleDownloadDialogFragment].
  */
 class DownloadsFeature(
     private val applicationContext: Context,
     var onNeedToRequestPermissions: OnNeedToRequestPermissions = { _, _ -> },
     var onDownloadCompleted: OnDownloadCompleted = { _, _ -> },
     private val downloadManager: DownloadManager = DownloadManager(applicationContext, onDownloadCompleted),
-    private val sessionManager: SessionManager
+    private val sessionManager: SessionManager,
+    private val fragmentManager: FragmentManager? = null,
+    private val dialog: DownloadDialogFragment = SimpleDownloadDialogFragment.newInstance()
```

Comment:
How close is `SimpleDownloadDialogFragment` to what we have in Focus? Can they just use this or will they need to do a lot of customization?

Reply:
They can't use this one because all customization that they are doing to the dialog. Their dialog should inherit from `DownloadDialogFragment` and if they use it with `DownloadsFeature` it will work automatically with minimal changes. In fact, I use focus as an example in the [docs](https://github.com/mozilla-mobile/android-components/blob/3d2a74127af0dac70483309cb931cbfedc4b1aa9/components/feature/downloads/README.md#downloaddialogfragment) 😄 .

```kotlin
class FocusDialogDownloadFragment : DownloadDialogFragment() {

    /*Creating a customized the dialog*/
    override fun onCreateDialog(bundle: Bundle?): AlertDialog {
        //DownloadsFeature will add these metadata before calling show() on the dialog.
        val fileName = arguments?.getString(KEY_FILE_NAME)
        //Not used, just for the sake you can use this metadata
        val url = arguments?.getString(KEY_URL)
        val contentLength = arguments?.getString(KEY_CONTENT_LENGTH)

        val builder = AlertDialog.Builder(requireContext())
        builder.setCancelable(true)
        builder.setTitle(getString(R.string.download_dialog_title))

        val inflater = activity!!.layoutInflater
        val dialogView = inflater.inflate(R.layout.download_dialog, null)
        builder.setView(dialogView)

        dialogView.download_dialog_icon.setImageResource(R.drawable.ic_download)
        dialogView.download_dialog_file_name.text = fileName
        dialogView.download_dialog_cancel.text = getString(R.string.download_dialog_action_cancel)
        dialogView.download_dialog_download.text =
            getString(R.string.download_dialog_action_download)

        dialogView.download_dialog_warning.text = getString(R.string.download_dialog_warning)

        setCancelButton(dialogView.download_dialog_cancel)
        setDownloadButton(dialogView.download_dialog_download)

        return builder.create()
    }

    private fun setDownloadButton(button: Button) {
        button.setOnClickListener {
            //Letting know DownloadFeature that can proceed with the download
            onStartDownload()
            TelemetryWrapper.downloadDialogDownloadEvent(true)
            dismiss()
        }
    }

    private fun setCancelButton(button: Button) {
        button.setOnClickListener {
           TelemetryWrapper.downloadDialogDownloadEvent(false)
            dismiss()
        }
    }
}

//Adding our dialog to DownloadsFeature
val downloadsFeature = DownloadsFeature(
            context(),
            sessionManager = sessionManager,
            fragmentManager = fragmentManager,
            dialog = FocusDialogDownloadFragment()
        )

downloadsFeature.start()
```